test/parallel/test-fs-promises.js

```diff
@@ -0,0 +1,165 @@
+    .then(common.mustNotCall())
+    .catch(common.expectsError({
+      code: 'ENOENT',
+      type: Error,
+      message:
+        'ENOENT: no such file or directory, access \'this file does not exist\''
+    }));
+}
+
+function verifyStatObject(stat) {
+  assert.strictEqual(typeof stat, 'object');
+  assert.strictEqual(typeof stat.dev, 'number');
+  assert.strictEqual(typeof stat.mode, 'number');
+}
+
+{
+  async function doTest() {
+    common.refreshTmpDir();
+    const dest = path.resolve(common.tmpDir, 'baz.js');
+    await copyFile(fixtures.path('baz.js'), dest);
+    await access(dest, 'r');
+
+    const handle = await open(dest, 'r+');
+    assert.strictEqual(typeof handle, 'object');
+
+    let stats = await fstat(handle);
+    verifyStatObject(stats);
+    assert.strictEqual(stats.size, 35);
+
+    await ftruncate(handle, 1);
+
+    stats = await fstat(handle);
+    verifyStatObject(stats);
+    assert.strictEqual(stats.size, 1);
+
+    stats = await stat(dest);
+    verifyStatObject(stats);
+
+    await fdatasync(handle);
+    await fsync(handle);
+
+    const buf = Buffer.from('hello world');
+
+    await write(handle, buf);
+
+    const ret = await read(handle, Buffer.alloc(11), 0, 11, 0);
+    assert.strictEqual(ret.bytesRead, 11);
+    assert.deepStrictEqual(ret.buffer, buf);
+
+    await chmod(dest, 0o666);
+    await fchmod(handle, 0o666);
+
+    await utimes(dest, new Date(), new Date());
+    await futimes(handle, new Date(), new Date());
+
+    await handle.close();
+
+    const newPath = path.resolve(common.tmpDir, 'baz2.js');
+    await rename(dest, newPath);
+    stats = await stat(newPath);
+    verifyStatObject(stats);
+
+    const newLink = path.resolve(common.tmpDir, 'baz3.js');
+    await symlink(newPath, newLink);
+
+    const newLink2 = path.resolve(common.tmpDir, 'baz4.js');
+    await link(newPath, newLink2);
+
+    stats = await lstat(newLink);
+    verifyStatObject(stats);
+
+    assert.strictEqual(newPath, await realpath(newLink));
+    assert.strictEqual(newPath, await readlink(newLink));
+
+    await unlink(newLink);
+    await unlink(newLink2);
+
+    // Verify unlink worked
+    try {
+      await stat(newPath);
+    } catch (err) {
+      common.expectsError({
+        code: 'ENOENT',
+        type: Error
+      })(err);
+    }
+
+    try {
+      await stat(newLink);
+    } catch (err) {
+      common.expectsError({
+        code: 'ENOENT',
+        type: Error
+      })(err);
+    }
+
+    const newdir = path.resolve(common.tmpDir, 'dir');
+    await mkdir(newdir);
+    stats = await stat(newdir);
+    assert(stats.isDirectory());
+
+    const list = await readdir(common.tmpDir);
+    assert.deepStrictEqual(list, ['baz2.js', 'dir']);
+
+    await rmdir(newdir);
+    try {
+      await stat(newdir);
+    } catch (err) {
+      common.expectsError({
+        code: 'ENOENT',
+        type: Error
+      })(err);
+    }
+
+    const base = path.resolve(common.tmpDir, 'FOO');
+    assert(
```

Comment:
Does this work on windows? `common.tmpDir` would contain `\` so it should not be able to be passed into `new RegExp` like that.

Reply:
heh, good point. will have to double check